# Chapter 3 - Ex1: Marketing Data
- Cho dữ liệu MarketingData.csv chứa số tiền chi tiêu hàng năm của 20.000 khách hàng của một công ty bán lẻ lớn. Nhóm tiếp thị của công ty đã sử dụng các kênh khác nhau để bán hàng hóa và đã phân loại khách hàng dựa trên các giao dịch mua được thực hiện bằng các kênh khác nhau, như sau: 0-Retail (Bán lẻ), 1-Road Show, 2-Social Media và 3-Televison. 
- Là phụ trách bộ phận Sale, bạn được giao nhiệm vụ xây dựng một mô hình Machine Learning có thể dự đoán kênh hiệu quả nhất có thể được sử dụng để nhắm mục tiêu khách hàng dựa trên chi tiêu hàng năm cho các sản phẩm (features) do công ty bán: Fresh (sản phẩm tươi sống), Milk(sữa), Grocery (tạp hóa), Frozen (sản phẩm đông lạnh), Detergents_Paper (chất tẩy rửa và giấy) và Delicassen (đồ ăn nhanh).

## Gợi ý

In [14]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split

from sklearn.metrics import classification_report,confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn import metrics
from sklearn.metrics import precision_recall_fscore_support

import matplotlib.pyplot as plt
import seaborn as sns

### Đọc dữ liệu, tiền xử lý dữ liệu

In [15]:
data= pd.read_csv('MarketingData.csv')
data.head(5)

Fresh         Milk      Grocery       Frozen  Detergents_Paper  \
0  6623.613537  5513.093240  6019.057354  5669.568008       5898.660607   
1  5642.542497  5829.866565  3960.339943  4270.020548       3498.818262   
2  5292.078175  6634.370556  4444.335138  4888.286021       3265.391352   
3  5595.227928  4754.860698  2977.856511  3462.490957       3609.264559   
4  5126.693267  6009.649079  3811.569943  4744.115976       3829.516831   

    Delicassen  Channel  
0  5179.234947        2  
1  4327.423268        2  
2  4887.560190        2  
3  4268.641413        0  
4  5097.491872        2

In [16]:
data.shape

(20000, 7)

In [17]:
data.isnull().values.any()

False

In [ ]:
data.describe()

In [18]:
data['Channel'].value_counts()

0    5007
3    5002
1    5001
2    4990
Name: Channel, dtype: int64

In [19]:
target = 'Channel'
X = data.drop(['Channel'],axis=1)
y=data[target]

In [20]:
X_train, X_test, y_train, y_test = train_test_split(X.values, y,
                                                    test_size=0.20, 
                                                    random_state=42)

In [21]:
from sklearn.linear_model import LogisticRegression

In [22]:
model = LogisticRegression(solver='saga',max_iter=300, multi_class='multinomial')

In [23]:
model.fit(X_train,y_train)

LogisticRegression(max_iter=300, multi_class='multinomial', solver='saga')

In [25]:
y_pred=model.predict(X_test)
y_proba = model.predict_proba(X_test)
print (y_proba)



[[4.32300187e-02 8.28882967e-01 1.27833055e-01 5.39589709e-05]
 [4.99929629e-03 1.78257902e-02 1.02323074e-01 8.74851839e-01]
 [8.18799259e-01 8.73343450e-02 9.35157021e-02 3.50694339e-04]
 ...
 [2.08768757e-02 1.97141982e-02 9.24312655e-03 9.50165800e-01]
 [7.86998233e-02 6.04227088e-01 1.80075321e-01 1.36997768e-01]
 [1.32704469e-01 4.01581949e-02 1.13851700e-02 8.15752166e-01]]


In [26]:
model.score(X_train, y_train)

0.6746875

In [27]:
model.score(X_test, y_test)

0.6795

In [28]:
precision_recall_fscore_support(y_test, y_pred, average='macro')

(0.6753444763299861, 0.678754939980805, 0.6764551768340966, None)

In [29]:
precision_recall_fscore_support(y_test, y_pred, average='micro')

(0.6795, 0.6795, 0.6795, None)

In [12]:
target_names = ["Retail","RoadShow","SocialMedia","Televison"]
print(classification_report(y_test, y_pred,target_names=target_names))

              precision    recall  f1-score   support

      Retail       0.68      0.65      0.67       982
    RoadShow       0.52      0.51      0.52       988
 SocialMedia       0.70      0.67      0.69      1027
   Televison       0.80      0.88      0.84      1003

    accuracy                           0.68      4000
   macro avg       0.68      0.68      0.68      4000
weighted avg       0.68      0.68      0.68      4000



In [13]:
cm = confusion_matrix(y_test, y_pred) 
cm

array([[642, 211,  76,  53],
       [180, 504, 176, 128],
       [ 83, 212, 693,  39],
       [ 34,  41,  48, 880]], dtype=int64)

In [ ]:
cm_df = pd.DataFrame(cm,
                     index = target_names, 
                     columns = target_names)
plt.figure(figsize=(8,6))
sns.heatmap(cm_df, annot=True,fmt='g',cmap='Blues')
plt.title('Logistic Regression \nAccuracy:{0:.3f}'.format(accuracy_score(y_test, 
                                                                         y_pred)))
plt.ylabel('True Values')
plt.xlabel('Predicted Values')
plt.show()

#### Nhận xét: 
- Model cho độ chính xác ~70% ----> chưa được cao
- Có giải pháp nào khác không? 
- Nếu chưa tìm được giải pháp nào phù hợp hơn thì có thể nghĩ đến việc phải thay đổi thuật toán (sẽ học sau)

##### Tuy nhiên, model trên đã giúp tìm ra được kênh hiệu quả nhất là  Televison với các thông số (0.80, 0.88, 0.84) đều cao